pkg/controller/keyvault/gateway_secret_provider_class.go

```diff
@@ -0,0 +1,249 @@
+)
+
+const (
+	saNameUriErr = "`App Routing requires a Certificate URI as well as a ServiceAccount that references an Identity that has access to your Keyvault in order to pull TLS certificates. Make sure both are specified in your Listener's TLS options for all listeners that require TLS certs."
+)
+
+// GatewaySecretProviderClassReconciler manages a SecretProviderClass for Gateway resource that specifies a ServiceAccount
+// and Keyvault URI in its TLS options field. The SPC is used to mirror the Keyvault values into
+// a k8s secret so that it can be used by the CRD controller.
+type GatewaySecretProviderClassReconciler struct {
+	client client.Client
+	events record.EventRecorder
+	config *config.Config
+}
+
+func NewGatewaySecretClassProviderReconciler(manager ctrl.Manager, conf *config.Config) error {
+	metrics.InitControllerMetrics(gatewaySecretProviderControllerName)
+
+	return gatewaySecretProviderControllerName.AddToController(
+		ctrl.
+			NewControllerManagedBy(manager).
+			For(&gatewayv1.Gateway{}), manager.GetLogger(),
+	).Complete(&GatewaySecretProviderClassReconciler{
+		client: manager.GetClient(),
+		events: manager.GetEventRecorderFor("aks-app-routing-operator"),
+		config: conf,
+	})
+}
+
+func (g *GatewaySecretProviderClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (res ctrl.Result, err error) {
+	res = ctrl.Result{}
+	err = nil
+
+	// set up metrics given result/error
+	defer func() {
+		metrics.HandleControllerReconcileMetrics(gatewaySecretProviderControllerName, res, err)
+	}()
+
+	// set up logger
+	logger, err := logr.FromContext(ctx)
+	if err != nil {
+		return
+	}
+	logger = gatewaySecretProviderControllerName.AddToLogger(logger).WithValues("name", req.Name, "namespace", req.Namespace)
+
+	// retrieve gateway resource from request + log the get attempt, but ignore not found
+	gwObj := &gatewayv1.Gateway{}
+	err = g.client.Get(ctx, req.NamespacedName, gwObj)
+
+	if err != nil {
+		err = client.IgnoreNotFound(err)
+		return
+	}
+
+	if gwObj.Spec.GatewayClassName != istioGatewayClassName {
+		return
+	}
+
+	// check its TLS options - needs to have both cert uri and either serviceaccount name or clientid
+	for index, listener := range gwObj.Spec.Listeners {
+		spc := &secv1.SecretProviderClass{
+			TypeMeta: metav1.TypeMeta{
+				APIVersion: "secrets-store.csi.x-k8s.io/v1",
+				Kind:       "SecretProviderClass",
+			},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      GenerateGwListenerCertName(gwObj.Name, listener.Name),
+				Namespace: req.Namespace,
+				Labels:    manifests.GetTopLevelLabels(),
+				OwnerReferences: []metav1.OwnerReference{{
+					APIVersion: gwObj.APIVersion,
+					Controller: util.ToPtr(true),
+					Kind:       gwObj.Kind,
+					Name:       gwObj.Name,
+					UID:        gwObj.UID,
+				}},
+			},
+		}
+		logger = logger.WithValues("spc", spc.Name)
+
+		if listenerIsKvEnabled(listener) {
```

Comment:
use go conventions of breaking early if it's not enabled instead of continously nesting

Reply:
updated